zerver/lib/bulk_create.py

```diff
@@ -145,3 +157,124 @@ def bulk_create_streams(realm: Realm,
+    reactions by emoji code and user profile and then return one reaction from
+    each group.
+    '''
+    for p in (prob_reaction, prob_repeat, prob_upvote):
+        if p > 1 or p < 0:
+            raise ValueError('Probability argument must be between 0 and 1.')
+    EMOJI_NAME = 0
+    EMOJI_CODE = 1
+    emojis = itertools.cycle(emojis)
+    if not users:
+        if message.recipient.type == Recipient.PERSONAL:
+            users = [
+                UserProfile.objects.get(recipient__id=message.recipient.id),
+                message.sender]
+        else:
+            users = [
+                subscription.user_profile
+                for subscription in Subscription.objects.filter(
+                    recipient=message.recipient)]
+        if not users:
+            return []
+    users = itertools.cycle(users)
+    reactions = []
+    if random.random() < prob_reaction:
+        emoji: Tuple[str, str] = next(emojis)
+        user: UserProfile = next(users)
+        reactions.append(Reaction(
+            user_profile=user,
+            message=message,
+            emoji_name=emoji[EMOJI_NAME],
+            emoji_code=emoji[EMOJI_CODE],
+            reaction_type=Reaction.UNICODE_EMOJI
+        ))
+        users_reacted = {user.id}
+        user = next(users)
+        while random.random() < prob_upvote and user.id not in users_reacted:
+            reactions.append(Reaction(
+                user_profile=user,
+                message=message,
+                emoji_name=emoji[EMOJI_NAME],
+                emoji_code=emoji[EMOJI_CODE],
+                reaction_type=Reaction.UNICODE_EMOJI
+            ))
+            users_reacted.add(user.id)
+            user = next(users)
+
+        while random.random() < prob_repeat:
+            emoji = next(emojis)
+            user = next(users)
+            reactions.append(Reaction(
+                user_profile=user,
+                message=message,
+                emoji_name=emoji[EMOJI_NAME],
+                emoji_code=emoji[EMOJI_CODE],
+                reaction_type=Reaction.UNICODE_EMOJI
+            ))
+            users_reacted = {user.id}
+            user = next(users)
+            while random.random() < prob_upvote and user.id not in users_reacted:
+                reactions.append(Reaction(
+                    user_profile=user,
+                    message=message,
+                    emoji_name=emoji[EMOJI_NAME],
+                    emoji_code=emoji[EMOJI_CODE],
+                    reaction_type=Reaction.UNICODE_EMOJI
+                ))
+                users_reacted.add(user.id)
+                user = next(users)
+    # try to avoid returning duplicated reactions
+    grouped_reactions = defaultdict(list)
+    for reaction in reactions:
+        k = (str(reaction.user_profile), str(reaction.emoji_code))
+        grouped_reactions[k].append(reaction)
+    return [reactions[0] for reactions in grouped_reactions.values()]
```

Comment:
Nice deduplication logic.  I'd use `reaction.user_profile_id` rather than `str()` on a Django model object; much cleaner with the same effect.